File: scripts/nodejs/RHELami.js
'use strict';
var fs = require('fs');
var __ = require('underscore'); 

/**
 * Computes RHEL amis for different instance types using AWS CLI.
 * AMIs are sorted by time and latest one is chosen.
 *
 *
 * @param {String} region
 * @param {String} RHEL Version
 * @returns {String} ami
 */

var child_process = require('child_process');
var argv = require('minimist')(process.argv.slice(2));
var version = argv['version'] || argv['v'] || "6.5";
var RHELami = 'RHEL-'+version + "*" + "-x86_64*";
var AWS = '/usr/local/bin/aws '

function escapeRegExp(string) {
    return string.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1");
}

function replaceAll(string, find, replace) {
  return string.replace(new RegExp(escapeRegExp(find), 'g'), replace);
}

function getRegions() {
  var cmd = AWS + " ec2 describe-regions"
  var regions = child_process.execSync(cmd, { encoding: 'utf8' });
  regions = JSON.parse(regions);
  return __.pluck(regions.Regions,'RegionName');
}

//array of regions
var regions =  argv['region'] && argv['region'].split(",") || getRegions();


function getAMIbyName(name,region,type) {
  var cmd = AWS + " ec2 describe-images --filters \"Name=name,Values=AMI-PLACEHOLDER\" \"Name=virtualization-type,Values=VTYPE-PLACEHOLDER\" --owners 309956199498 --region REGION-PLACEHOLDER";
  cmd = replaceAll(cmd,"AMI-PLACEHOLDER",name);
  cmd = replaceAll(cmd,"REGION-PLACEHOLDER",region);
  cmd = replaceAll(cmd,"VTYPE-PLACEHOLDER",type);

  var ami = child_process.execSync(cmd, { encoding: 'utf8' });

  return JSON.parse(ami);
}

function CompareCreationDate(a,b) {
	return Date.parse(a.CreationDate) - Date.parse(b.CreationDate) 
}

function getRHELamis() {
  var types = ["paravirtual","hvm"];
  var amis = {};
  for (var v in regions) {
  	var region = regions[v];
  	var data = {};
  	for (var t in types) {
        try {
            var type = types[t];
            var ami = getAMIbyName(RHELami,region,type);
            var amiArray = Object.keys(ami.Images).map(function (key) {return ami.Images[key]});
            var amiArraySorted = amiArray.sort(CompareCreationDate);
            data[type] = amiArraySorted[amiArraySorted.length-1].ImageId;            
        }
        catch(e) {
            data[type] = ''
        }
  	}
  	amis[region] = data;
  }
  return amis;
}

var RHELamis = getRHELamis();

console.log(JSON.stringify(RHELamis, null, 4));

File: scripts/RHELami.py
import argparse
import subprocess
import json
import dateutil.parser

aws_cmd = '/usr/local/bin/aws '

def exe_cmd(cmd,cwd=None):
    if cwd == None:
        proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, shell=True)
        proc.wait()
        (out, err) = proc.communicate()
        output = {}
        output['out'] = out
        output['err'] = err
        return output
    else:
        proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, shell=True,cwd=cwd)
        proc.wait()
        (out, err) = proc.communicate()
        output = {}
        output['out'] = out
        output['err'] = err
        return output

def get_creationdate(data):
    return  dateutil.parser.parse(data['CreationDate'])


def main():
    parser = argparse.ArgumentParser(description='Find latest RHEL AMI')
    parser.add_argument('-v', dest="version",metavar="VERSION",required = True,
                              help='RHEL Version')
    parser.add_argument('-r', dest="region",metavar="REGION",required = True,
                              help='REGION to query AMI')
    parser.add_argument('-t', dest="type",metavar="AMI_TYPE",required = True,
                              help='AMI Type (hvm)')

    args = parser.parse_args()
    version = args.version
    region = args.region
    ami_type = args.type

    ami_name = 'RHEL-'+ version + "*" + "-x86_64*";

    cmd = aws_cmd + " ec2 describe-images --filters \"Name=name,Values=AMI-PLACEHOLDER\" \"Name=virtualization-type,Values=VTYPE-PLACEHOLDER\" --owners 309956199498 --region REGION-PLACEHOLDER"
    cmd = cmd.replace('AMI-PLACEHOLDER',ami_name)
    cmd = cmd.replace('VTYPE-PLACEHOLDER',ami_type)
    cmd = cmd.replace('REGION-PLACEHOLDER',region)

    output = exe_cmd(cmd)
    images = output['out']
    val = json.loads(images)
    images =  val['Images']
    sorted_images = sorted(images,key = get_creationdate,reverse=True)
    print sorted_images[0]['ImageId']



if __name__ == "__main__":
    main()
